### ISSUE TITLE
Multi - window

### ISSUE THREAD

**Issue by Denys2211** [WPF-Pilot/WpfPilot#7]
Exception after opening one more window. Here 
  GetResponse -> TimeoutAfter

**Method:**
 AppDriver.GetElements(x => x["Name"] == "OtherWindow").Click()

**Message:** 
   "The operation has timed out."

**Stack Tracy:** 
   at WpfPilot.Interop.NamedPipeClient.<TimeoutAfter>d__4`1.MoveNext() in 
                  C:\code\WpfPilot\WpfPilot\Interop\NamedPipeClient.cs:line 124
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in
                  System.Runtime.ExceptionServices\ExceptionDispatchInfo.cs:line 72
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) in 
                  System.Runtime.CompilerServices\TaskAwaiter.cs:line 88
   at WpfPilot.Interop.NamedPipeClient.<>c__DisplayClass2_0.<GetResponse>b__0() in 
                  C:\code\WpfPilot\WpfPilot\Interop\NamedPipeClient.cs:line 85
   at WpfPilot.Utility.Retry.With(Action action, Int32 retryIntervalMs, Int32 retryCount) in
                  C:\code\WpfPilot\WpfPilot\Utility\Retry.cs:line 16

Environment:

.Net Framework 4.8
x64
WpfPilot v1.0.2

**Comment by ELanning** [comment]
Interesting, I had briefly tested with multi windows before, but nothing robust. I’ll take a look. Thanks for the report.

**Comment by ELanning** [comment]
@Denys2211 I'm having a hard time reproducing this issue. This added [regression test ](https://github.com/WPF-Pilot/WpfPilot/pull/8/files#diff-96274a78983cc95d5aadce42e3c1c7a01b1e7aeb69f028c383b402f1d29e7904R73) passed on net framework and dot net.

Are you setting up windows differently? Thanks for any extra clarity.

**Comment by Denys2211** [comment]
I see. I get the exception when I try open one more window after button click.

repro steps:

1. open MainWindow
2. click on  button for show other window


**event handler:**

```
private void OtherWindow_Click(object sender, RoutedEventArgs e)
{
	OtherWindow otherWindow = new OtherWindow();
	otherWindow.Show();
}
```

**Comment by ELanning** [comment]
Denys, I see your original code snippet:

>AppDriver.GetElements(x => x["Name"] == "OtherWindow").Click()

Are you clicking the window to try and bring it into focus? Would this code snippet below work for you instead?

> appDriver.GetElement(x => x["Title"] == "OtherWindow").Invoke\<Window\>(x => x.Activate());

I also see you're using `Name` when most Windows probably use `Title`. Could you post the Window xaml? Here is mine:

```
<Window x:Class="WpfPilot.ExampleApp.OtherWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:WpfPilot.ExampleApp"
        mc:Ignorable="d"
        Title="OtherWindow" Height="450" Width="800">
```

I updated my PR to try and get a repro and still couldn't, so I think the issue might be elsewhere.

**Comment by Denys2211** [comment]
AppDriver.GetElements(x => x["Name"] == "OtherWindow").Click() -  "OtherWindow" it is just buttons name.

when I click on the button I get exception.  By the button I open new window. Window opened successfully but in the end click method I get error.

I see you put Show method for OtherWindow in MainWindow constructor but for reproduce the error you need put this in  click event handler

**Comment by ELanning** [comment]
Does your click handler take more than 10 seconds to execute or does it do async work? I think that might be the issue and I could fix it by allowing longer time.

> I see you put Show method for OtherWindow in MainWindow constructor but for reproduce the error you need put this in click event handler

Yea I changed it to the click event handler [here](https://github.com/WPF-Pilot/WpfPilot/pull/8/files#diff-d1a73368f378617766505c38e2e2d77106c106324ba3e6c52c8b19dac7bbc63bR93), still seemed to work

**Comment by Denys2211** [comment]
Doesn't, in original it opened fast - 1 second or less

**Comment by ELanning** [comment]
Well I'd love to get this fixed. If you could write a minimal reproducible example and post a Github link to it, I'd pay you for your time.

Other than that, I'm out of ideas at the moment, but I'll post here if I think of anything.

**Comment by Denys2211** [comment]
Yes, I just checked and I see with empty new window it works well

**Comment by Denys2211** [comment]
I found reason  - use ShowDialog instead Show

```
private void OtherWindow_Click(object sender, RoutedEventArgs e)
{
	OtherWindow otherWindow = new OtherWindow();
	otherWindow.ShowDialog();
}
```

with ShowDialog  method you can reproduce the exception. Sorry I didn't check it before.


if I use ShowDialog method I get Task.Status = WaitingForActivation (completedTask != task). It gives me throw:

```
private static async Task<TResult> TimeoutAfter<TResult>(Task<TResult> task, TimeSpan timeout)
{
	using var timeoutCancellationTokenSource = new CancellationTokenSource();
	var completedTask = await Task.WhenAny(task, Task.Delay(timeout, timeoutCancellationTokenSource.Token));
	if (completedTask == task)
	{
		timeoutCancellationTokenSource.Cancel();
		return await task; // Very important in order to propagate exceptions
	}
	else
	{
		throw new TimeoutException("The operation has timed out.");
	}
}
```

**Comment by ELanning** [comment]
Thank you, that probably explains it. I’ll take a look today. Kindly let me know if I owe you anything for that investigation.

**Comment by ELanning** [comment]
Fixed with a hacky workaround [here](https://github.com/WPF-Pilot/WpfPilot/pull/8).

Should be in version 1.0.3 after it's verified and released on NuGet in a few hours.

**Comment by Denys2211** [comment]
Unfortunately it doesn't work as expected, I always get false from ShowDialog method.  
Yes, replace `ShowDialog` with `Show` this is unideal.

**Comment by ELanning** [comment]
Ah I will have it return true and leave it in place until I think of a better solution. A workaround on users end is detecting if the app is being run in test mode, and adjusting various code accordingly.

Eg:
`if (testmode) { prefill fake values } else { showDialog }`

EDIT: Changed in [v 1.0.4](https://github.com/WPF-Pilot/WpfPilot/commit/179e9d73a602294d20d60e7b2c12319fd1a49860#diff-146686f23222dce50edb04fa6586199f597906155d137600545277dc37f041faR218)

**Comment by Denys2211** [comment]
Thank you for your time, your fixes, but I need get variable result - true or false and after users activity. Unfortunately, I can't use it for my app.

**Comment by ELanning** [comment]
Yea, I will likely be able to come up with a better fix when I have the time. Sorry about that.

**Comment by ELanning** [comment]
CC @Denys2211 Fixed in version 1.1.0. I am assuming you have already moved on, but I feel obligated to at least alert you.

Sorry I couldn't provide a more robust fix sooner!
